### ISSUE TITLE
Initializer

### ISSUE THREAD

**Issue by SammyROCK** [Cappta/best-practices#33]
Se houver uma exceção em alguma atribuição do Initializer, ela será lançada na primeira linha do bloco de código, não especificando onde ocorreu o problema.
Além disso, em conjunto com o posicionamento das chaves, torna o código mais longo.

Portanto temos 2 malefícios e nenhum beneficio.

Por esses motivos, sou contra a utilização de Initializers, independente do cenário.

**Comment by vinibeloni** [comment]
Garantir que os valores são validos não é de responsabilidade do objeto que recebe os valores e sim que passa, ao dizer que:

> Se houver uma exceção em alguma atribuição do Initializer, ela será lançada na primeira linha do bloco de código, não especificando onde ocorreu o problema

Pra mim,isso significa que em alguns casos não houve um tratamento adequado já que se por algum motivo um valor inválido for atribuído no objeto e por não ser tratado só depois de seila quantas linhas/métodos depois vai estourar esse cara se o valor atribuído não válido. Além de sujar mais o código também dependendo do caso, se fosse pensar em:

```csharp

Student student = new Student();
student.StudentID = 1;
student.StudentName = "Bill";
student.Age = 20;
student.StandardID = 10;
student.Address = "Test";

//ou
Student  student = new Student
{
  StudentID = 1,
  StudentName = "Bill",
  Age = 20,
  StandardID = 10,
  Address = "Test"
};
```

> em conjunto com o posicionamento das chaves, torna o código mais longo.

O valor agregado é legibilidade. Nossos IFs também usam chave e mesmo que o valor dela seja evitar que passe despercebido, aumenta o código também. 

**Então pra mim**, o trade-off é levemente ruim, por causa de legibilidade e de certa forma duplicação como no exemplo acima.

Então resumindo, **na minha opinião**, é claro que aqui serve para discutir, mas se estourou no initilializer, é ausencia de uma validação e também não vejo como um grande problema ele me mostrar na mesma linha do objeto e não na linha da propriedade ou algo assim, isso já significa que algo não está correto

Conjunto de chaves também não vejo problemas, já que se for uma propriedade obrigatória para aquele objeto, deveria estar no construtor da classe.

**Comment by SammyROCK** [comment]
"em alguns casos não houve um tratamento adequado"
Bem vindo à realidade, caso contrario não haveria problema.

"O valor agregado é legibilidade."
2 linhas e indentação vs 1 bloco de código
Além disso, na maioria dos casos não será com valores constantes e pode ser facilmente criado um construtor:
`Cachorro cachorro = new Cachorro(arroz, feijao);`
E, mesmo se for, da para isolar em uma função.
`Student CreateStudent()
{
var student = new Student();
student.StudentID = 1;
student.StudentName = "Bill";
student.Age = 20;
student.StandardID = 10;
student.Address = "Test";
return student;
}

var student = this.CreateStudent();`

Discordo com a comparação do IF, pois para o IF não é opcional a criação de um novo bloco e não ofusca problemas contidos nesse bloco.

Se estourou no Initializer, ao invés de corrigir o problema, vai caçar doninha em todas as possibilidades e gerar uma versão com mais informação para só depois que o cenário for replicado conseguir corrigir?

Acho um péssimo trade-off para um beneficio questionável, mesmo em condições ideais.

**Comment by vinibeloni** [comment]
> "em alguns casos não houve um tratamento adequado"
Bem vindo à realidade, caso contrario não haveria problema.

Justamente, se existe um problema você arruma, por isso a validação, senão não faz sentido nenhum validar qualquer coisa. Eu não pagaria pra ver um cara estourando em outro lado do código por causa de 2 chaves. 

> 2 linhas e indentação vs 1 bloco de código

Resolvido:
```csharp
var cachorro = new Cachorro() { Arroz = arroz }; 
```

> Se estourou no Initializer

Pra mim isso não deveria acontecer em hipótese alguma, como que eu mando um valor que  está sobre meu controle pra um objeto e fazer ele estourar la dentro? Se os valores estão corretos e eu to garantindo que, no minimo, não vão quebrar minha aplicação, por que eu vou colocar separados linha por linha? se eu tiver um valor que quebra minha aplicação, de forma alguma eu devo permitir que ele passe.

O ponto das chaves a mais, são 2 chaves com ocupando uma coluna e uma linha
com a atribuição linha por linha, são N colunas ocupadas mais a sua propria linha.

Acho o trade-off válido já que teoricamente devemos validar qualquer valor antes de passar pra frente dentro das minhas regras de négocio.

Se for o caso, podemos fazer uma reunião sem problema algum pra discutir esta issue e se vale a pena ou não determinar isso como algo obrigatório. De verdade, tem coisas que eu não acho que deveriam ser colocadas em uma convenção, como está que estamos falando já que é algo mais de personalidade de quem escreve, maaas, temos todo o time pra resolver isso de forma mais efetiva.

**Comment by SammyROCK** [comment]
`var cachorro = new Cachorro() { Feijao = arroz.Feijao, Goiaba = arroz.Goiaba, Gato = arroz.Gato };`
Resolvido?

Você está cometendo o erro básico de achar que seu código é perfeito e não haverá falhas, mesmo se outra pessoa fizer alguma alteração.
Sua opinião se baseia na premissa que o código não haverá erros, pois tudo está perfeitamente alinhado e testado no melhor momento. A realidade não é essa.

**Comment by daniellessio** [comment]
Exceção estourar na primeira linha ou na ultima, dificulta o debug. Mas ter caso de teste cobrindo entrada + validações do código que esta inicializando o objeto em si facilita a vida, e esse comportamento defensivo (que me parece padrão) mitiga a sua preocupação @SammyROCK.
Por isso prefiro a legibilidade apontada pelo @vinibeloni e concordo que o código fica mais limpo com o Initializer, mesmo que mais longo. Inclusive, nesse ponto prefiro a posição das chaves que é padrão no JS e que estou começando a ver em projetos .NET/ASP.NET Core.

Acho que é um bom ponto para levarmos para discussão de todo mundo e promover algum bate papo voltado para padrão. O que acham?

**Comment by vinibeloni** [comment]
@daniellessio concordo sim e até sugiro que isso aconteça se está discussão acabar se estendendo muito.

Só queria fazer um adendo, que pra mim "erro básico" é fazer um atributo explodir um erro por falta de validação do valor que ele recebeu.

Você está se baseando que tudo que passa pelo seu código vai explodir a qualquer momento, eu só parti do ponto que se foi feito as validações corretas na entrada você ameniza o risco de sofrer com problemas. 
Pra mim não faz diferença se a linha 1 ou 2 aconteceu o erro, pra mim é porque o cara que ocasionou esse erro, deixou isso passar. Isso não é o código perfeito, é fazer o correto. 

Imagina se eu pego o valor de uma config, aí ao inves de validar se é um Int32 ou se existe essa chave, eu faço isso:

```csharp
var cachorro = new Cachorro(){ Raça = ConfigurationManager.AppSettings["Raça"].ToString() };

var cachorro = new Cachorro();
cachorro.Raça = ConfigurationManager.AppSettings["Raça"].ToString();
```
os dois vão dar erro, e o problema não é nem um nem outro. O problema é: não validei se existe a chave e se o valor é o que eu espero

então estourar aqui ou lá, tanto faz. Eu não me garanti em nenhum dos dois casos. Então o argumento que object initializer é ruim por causa disso, nem relação eu vejo um com o outro.

**Comment by valescaf** [comment]
encapsulamento: https://pt.wikipedia.org/wiki/Orienta%C3%A7%C3%A3o_a_objetos

**Comment by SammyROCK** [comment]
No mundo ideal o problema não aconteceria.
Mas no mundo real, código perfeito é utopia. Vão surgir demandas, urgências e até mesmo dias em que a pessoa não está em seu melhor e vão passar problemas despercebidos.
Usar um padrão que foge da realidade não faz sentido, ignorar o efeito colateral por preciosismo tecnológico não pode acontecer.

Além disso o Initializer encoraja duplicar código, pois ao invés de fazer um construtor ou um método que realiza a instanciação, será utilizada uma instanciação genérica e duplicado esse código onde ele for usado.
